eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,365 @@
+	}
+	res, err := mc.call(contextWithMethod(ctx, "SubscribeFilterLogs"), f)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.(ethereum.Subscription), nil
+}
+
+func (mc *MultiClient) FilterLogs(ctx context.Context, q ethereum.FilterQuery) ([]ethtypes.Log, error) {
+	f := func(client SingleClientProvider) (any, error) {
+		return client.FilterLogs(ctx, q)
+	}
+	res, err := mc.call(contextWithMethod(ctx, "FilterLogs"), f)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.([]ethtypes.Log), nil
+}
+
+func (mc *MultiClient) Filterer() (*contract.ContractFilterer, error) {
+	return contract.NewContractFilterer(mc.contractAddress, mc)
+}
+
+func (mc *MultiClient) ChainID(_ context.Context) (*big.Int, error) {
+	return mc.chainID, nil
+}
+
+func (mc *MultiClient) Close() error {
+	close(mc.closed)
+
+	var multiErr error
+	for i, client := range mc.clients {
+		if err := client.Close(); err != nil {
+			mc.logger.Debug("Failed to close client", zap.String("address", mc.nodeAddrs[i]), zap.Error(err))
+			multiErr = errors.Join(multiErr, err)
+		}
+	}
+
+	return multiErr
+}
+
+func (mc *MultiClient) call(ctx context.Context, f func(client SingleClientProvider) (any, error)) (any, error) {
+	if len(mc.clients) == 1 {
+		return f(mc.clients[0])
+	}
+
+	// Iterate over the clients in round-robin fashion,
+	// starting from the most likely healthy client (currentClientIndex).
+	var startingIndex = int(mc.currentClientIndex.Load())
+	var allErrs error
+	for i := 0; i < len(mc.clients); i++ {
+		clientIndex := (startingIndex + i) % len(mc.clients)
+		nextClientIndex := (clientIndex + 1) % len(mc.clients) // For logging.
+		client := mc.clients[clientIndex]
+
+		logger := mc.logger.With(
+			zap.String("addr", mc.nodeAddrs[clientIndex]),
+			zap.String("method", methodFromContext(ctx)))
+
+		// Make sure this client is healthy. This shouldn't cause too many requests because the result is cached.
+		// TODO: Make sure the allowed tolerance doesn't cause issues in log streaming.
+		if err := client.Healthy(ctx); err != nil {
+			logger.Warn("client is not healthy, switching to next client",
+				zap.String("next_addr", mc.nodeAddrs[nextClientIndex]),
+				zap.Error(err))
+			allErrs = errors.Join(allErrs, err)
+			mc.currentClientIndex.Store(int64(nextClientIndex)) // Advance.
+			continue
+		}
+
+		logger.Debug("calling client")
+
+		v, err := f(client)
+		if errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled) {
+			mc.logger.Debug("received graceful closure from client", zap.Error(err))
+			return v, err
+		}
+
+		if err != nil {
+			mc.logger.Error("call failed, trying another client",
+				zap.String("next_addr", mc.nodeAddrs[nextClientIndex]),
+				zap.Error(err))
+
+			allErrs = errors.Join(allErrs, err)
+			mc.currentClientIndex.Store(int64(nextClientIndex)) // Advance.
+			continue
+		}
+
+		// Update currentClientIndex to the successful client.
+		mc.currentClientIndex.Store(int64(clientIndex))
```

Comment:
Probably, there could be a case in which a node fails on some of the requests (e.g. some API subset is disabled), so the clients would be switched back and forth. But I'm not sure we need to do something with it now